### PR TITLE
Expose TXT registers to securityfs (slaunch directory)

### PR DESCRIPTION
**Pull request by garnetgrimm**
Exposed STS,ESTS,ERRORCODE,DIDVID,VER.EMIF,SCRATCHPAD and E2STS registers

### CODE REVIEW

File: arch/x86/kernel/slaunch.c [hunk 1]
```diff
@@ -27,13 +27,72 @@
...
+static char msg_buffer[MSG_BUFFER_LEN];
+
 static u32 sl_flags;
 static struct sl_ap_wake_info ap_wake_info;
 
+static ssize_t reg_write(struct file *file, const char __user *buf, size_t datalen, loff_t *ppos) {
+	return -EINVAL;
+}
+
+static void txt_info_to_buffer(unsigned int offset, int size) {
```
Comment (dpsmith) [inline arch/x86/kernel/slaunch.c:58]: You are passing the result of `sizeof` as the size param and `sizeof` is of type size_t which is unsigned thus making an implicit cast to a signed int. This then results in the comparison between a signed int and unsigned int in your switch statement later.

File: arch/x86/kernel/slaunch.c [hunk 2]
```diff
@@ -27,13 +27,72 @@
...
 
+#define MSG_BUFFER_LEN 20
+#define MSG_ERROR "Mapping Error\n"
+
 #define PREFIX	"SLAUNCH: "
 
+#define DECLARE_PUB_READ(reg_name, reg_offset, reg_size)					\
+static ssize_t reg_name##_read(struct file *flip, char *buffer, size_t len, loff_t *offset) {	\
+	txt_info_to_buffer(reg_offset, reg_size);						\
+	return simple_read_from_buffer(buffer, len, offset, &msg_buffer, MSG_BUFFER_LEN);	\
```
Comment (dpsmith) [inline arch/x86/kernel/slaunch.c:41]: Why are you bouncing through a static buffer instead of just writing the info directly into the destination buffer?

File: arch/x86/kernel/slaunch.c [hunk 3]
```diff
@@ -27,13 +27,72 @@
...
+}												\
+static const struct file_operations reg_name##_ops = {						\
+	.read = reg_name##_read,								\
+	.write = reg_write									\
```
Comment (dpsmith) [inline arch/x86/kernel/slaunch.c:45]: I am pretty sure you don't need to define/register a write handler.

File: arch/x86/kernel/slaunch.c [hunk 4]
```diff
@@ -27,13 +27,72 @@
...
 #include <asm/setup.h>
+#include <linux/fs.h>
 #include <linux/slaunch.h>
 
+#define MSG_BUFFER_LEN 20
+#define MSG_ERROR "Mapping Error\n"
+
 #define PREFIX	"SLAUNCH: "
 
+#define DECLARE_PUB_READ(reg_name, reg_offset, reg_size)					\
```
Comment (dpsmith) [inline arch/x86/kernel/slaunch.c:38]: Please move all this down to be colocated with the rest of the securityfs code.

File: arch/x86/kernel/slaunch.c [hunk 5]
```diff
@@ -553,23 +624,39 @@ static long slaunch_expose_securityfs(void)
...
-					   &sl_evtlog_ops);
-		if (IS_ERR(fs_entries[entry])) {
-			pr_err("Error creating securityfs %s file\n",
-				sl_evtlog.name);
-			ret = PTR_ERR(fs_entries[entry]);
+		
+		ret = sl_create_file(--entry, sl_evtlog.name, &sl_evtlog_ops);
+		if (ret)
+			goto err_dir;
+		ret = sl_create_file(--entry, "txt_sts",&sts_ops);
```
Comment (dpsmith) [inline arch/x86/kernel/slaunch.c:631]: These entries should be under a separate test for `(sl_flags & SL_FLAG_ARCH_TXT)`. It would also be more logical instead of having to prefix all the nodes with `txt_`, to create a txt directory node and place these entries under that node.

File: include/linux/slaunch.h [hunk 6]
```diff
@@ -66,12 +66,14 @@
 #define TXT_CR_CMD_RESET		0x0038
 #define TXT_CR_CMD_CLOSE_PRIVATE	0x0048
 #define TXT_CR_DIDVID			0x0110
+#define TXT_CR_VER_EMIF			0x0200
```
Comment (dpsmith) [inline include/linux/slaunch.h:69]: This file is introduced in an earlier patch commit, it would be cleaner to separate this into a separate commit that is tagged to be merged with the patch commit this is introduced in.

**Comment by dpsmith** [comment]
I have give a first pass on the changes. I would suggest first to review the approach[1] we are using to track changes to ease consolidating all change into a clean patch series for sending to LKML. You will need to adjust your commit messages to reflect the commit the pieces will be merged with and add a signoff-by line that will get included in the patches your commits will be merged to ensure compliance with Linux's guidelines[2]. Along those lines, please add yourself to the Authors section at the top of the file where you made significant logic changes.

[1] https://groups.google.com/forum/#!topic/trenchboot-devel/AxomiCh6rrs
[2] https://www.kernel.org/doc/html/v4.17/process/submitting-patches.html#sign-your-work-the-developer-s-certificate-of-origin

File: arch/x86/kernel/slaunch.c [hunk 7]
```diff
@@ -27,13 +27,72 @@
 #include <asm/e820/api.h>
 #include <asm/bootparam.h>
 #include <asm/setup.h>
+#include <linux/fs.h>
```
Comment (rossphilipson) [inline arch/x86/kernel/slaunch.c:30]: This should go in a separate patch to make our rebases of the patch set go as smoothly as they can.

File: arch/x86/kernel/slaunch.c (hunk 5 above)
Comment (dpsmith) [reply arch/x86/kernel/slaunch.c:631]: Just to be clear, the addition of these entries should be independent/outside of the check if the log addr has been set and toggled with the test in the initial comment.

File: arch/x86/kernel/slaunch.c (hunk 7 above)
Comment (rossphilipson) [reply arch/x86/kernel/slaunch.c:30]: Specifically this on line include should go with 2ba13df24ceca280eec1e42a3fbdc60a1f3fc4c8.

File: arch/x86/kernel/slaunch.c (hunk 4 above)
Comment (rossphilipson) [reply arch/x86/kernel/slaunch.c:38]: Once you co-locate all your securityfs related changes, they will go with commit 5544017b55129000fb8ada14c09a7cfcfc47be62

File: include/linux/slaunch.h (hunk 6 above)
Comment (rossphilipson) [reply include/linux/slaunch.h:69]: Yes, these additions should go with commit ba2f29985aba9dc09518a9130c60fc0bc1066b0a

File: arch/x86/kernel/slaunch.c (hunk 2 above)
Comment (garnetgrimm) [reply arch/x86/kernel/slaunch.c]: In my latest commit I changed msg_buffer to be a local variable. I tried to get rid of it entirely but without it I don't believe there is any way to properly format the data with sprintf.

File: arch/x86/kernel/slaunch.c (hunk 7 above)
Comment (garnetgrimm) [reply arch/x86/kernel/slaunch.c]: linux/fs is no longer needed so I just removed this line

File: arch/x86/kernel/slaunch.c (hunk 2 above)
Comment (dpsmith) [reply arch/x86/kernel/slaunch.c]: Thinking about it, all of these are individual values not a long buffer that could require multiple reads from user space. You should just set them up using DEVICE_ATTR and then you can register them as a device attr group which will automatically take care of the subdirectory concern.

File: arch/x86/kernel/slaunch.c [hunk 8]
```diff
@@ -5,7 +5,7 @@
  *
  * Author(s):
  *     Daniel P. Smith <dpsmith@apertussolutions.com>
- *
+ *     Garnet T. Grimm <garnetgrimm720@gmail.com>
```
Comment (dpsmith) [inline arch/x86/kernel/slaunch.c:8]: Your Author and Signed-off need to be by the same email address.

File: arch/x86/kernel/slaunch.c (hunk 7 above)
Comment (rossphilipson) [reply arch/x86/kernel/slaunch.c]: ack

File: arch/x86/kernel/slaunch.c (hunk 2 above)
Comment (garnetgrimm) [reply arch/x86/kernel/slaunch.c]: hm, would I be instead exposing the registers over devfs if I did that? I'm not sure how to use DEVICE_ATTR or .show with securityfs.

File: arch/x86/kernel/slaunch.c (hunk 8 above)
Comment (garnetgrimm) [reply arch/x86/kernel/slaunch.c]: fixed

File: arch/x86/kernel/slaunch.c [hunk 9]
```diff
@@ -466,6 +466,57 @@ void __init slaunch_setup(void)
...
 }
 
+#define MSG_BUFFER_LEN 20
+#define MSG_ERROR "Mapping Error\n"
+
+#define DECLARE_PUB_READ(reg_name, reg_offset, reg_size)					\
+static ssize_t reg_name##_read(struct file *flip, char *buffer, size_t len, loff_t *offset) {	\
+	char msg_buffer[MSG_BUFFER_LEN];							\
+	txt_info_to_buffer(reg_offset, reg_size, msg_buffer);					\
+	printk(KERN_INFO "%s", msg_buffer);							\
```
Comment (rossphilipson) [inline arch/x86/kernel/slaunch.c:476]: checkpatch is going to complain about using printk. Use something like pr_info instead.

Comment (garnetgrimm) [reply arch/x86/kernel/slaunch.c]: fixed - just removed the print statement entirely

File: arch/x86/kernel/slaunch.c (hunk 2 above)
Comment (dpsmith) [reply arch/x86/kernel/slaunch.c]: Ack, I got crossed with sysfs.

**Comment by tsirakisn** [comment]
@dpsmith @rossphilipson @eric-ch thanks all for your time with this :+1:
